guides/how-to-decaf.md

```diff
@@ -0,0 +1,79 @@
+
+## What is Decaffeination?
+
+The process of **Decaffeinating (decaf)** a codebase, is the part automated, part manual task of turning valid CoffeeScript code into valid and functionally identical JavaScript code.
+
+## Who is this for and what will you learn?
+
+This guide is made for any new contributors to Pulsar, or existing contributors who have never decaffeinated code, that want to help in the projects goals of converting all of our CoffeeScript code to JavaScript.
+
+This guide should teach you the following:
+
+  * Know what Decaffeination is.
+  * Best Practices to decaf code.
+  * Best Practices to decaf code for Pulsar.
+  * How to make your decaf work easy to review for maintainers.
+
+## Overview
+
+This section covers the general overview and workflow you should adopt to successfully decaf some code.
+
+But first to do so, there is one big rule:
+
+  * No matter what, any PR that decafs **both** the source files of a repo/package **and** the specs cannot be accepted. Since decaffing code can be tricky, we need to ensure we catch any errors that are being produced. If both the specs are decaffed at the same time the source files are, it is impossible to tell if the process happened successfully, and which aspect is producing any test failures or errors we see later on. Generally, if a repo/package needs both the specs and source files decaffed, they should each occur as their own unique PRs.
+
+Secondly, here is a general recommendation when choosing what to decaf:
+
+  * When you decide to decaf a repo/package, it's a good idea to decaf the entirety of what you are working on. Decaffeinating only a single spec file or single source file doesn't do us much good, and in many cases may be totally invalid on it's own, so it's highly recommended to, within a single PR, decaf the entire source files, or entire spec files.
+
+With those rules/recommendations out of the way, following below is the general guide that should be followed:
+
+1. As you might expect, decide what you are going to decaffeinate. To help make everyone's lives easier, you should ensure the file you are decaffing doesn't have any open PRs against it, or be at risk of receiving major changes anytime soon. If you aren't sure of this, feel free to ask the Pulsar maintainers.
+
+2. Once you've decided what repo/package you are going to decaf, then decide if you want to decaf the specs or source files.
+
+3. Now with the choices out of the way, you can get started. You'll want to make sure to create the Machine Decaf branch (more on that later) but this would be a branch that exists on the remote repo that contains only machine decaffed files. (If you don't have permissions to make this, feel free to ask the Pulsar maintainers). While we could be done there, the machine decaffed code is valid, but very unreadable and difficult to work with, so it requires manual intervention to improve.
```

Comment:
This seems confusing to me, are we saying that the machine decaf branch exists on the original repo before one can start? This seems somewhat difficult for people to get involved with.

Reply:
While I know that's a hard barrier to entry it does. 
The remote repository needs to have a machine decaf branch first, so that the contributor doing the manual decaf work is able to create a branch from those changes (so they have the machine decaf), and more importantly, it's the branch they target to merge their manual decaf. 

Since, there's no way to target a branch that doesn't exist on remote in a PR that's why it needs to already exist. Since if we don't have a machine decaf and manual decaf diff, these are impossible to review.

That's why I mentioned that if the contributor doesn't have permissions that they can contact a Pulsar maintainer to have that branch made, as this is how I've been doing all my decaf work, and how it's being done for PPM. Before the latest PPM decaf PRs were made I had created a machine decaf branch for the contributor to target.

Reply:
Ok I get you, I was thinking that an alternative is to submit a PR to create a machine decaf branch as a separate PR from the manual work but thinking back I realise that isn't possible without having access first as you can only compare branches in a PR, not target the creation of a new one.

Reply:
Yeah, I wish there was a better way to do this.

We could possibly instruct to create the machine decaf as a single commit, then only commits afterwards can add manual decaf steps. And we review a PR based on the diff between commits, but that sounds like it's much more likely to be done improperly, but even so not sure that's worth the added ability of anyone being able to work on it, without having to wait on us.

Reply:
I agree, I think what you suggest is probably best.